Issues/1249 support prepare in shaclsail

hmottestad opened the pull request:
This PR addresses GitHub issue: eclipse/rdf4j#1249 .

Briefly describe the changes proposed in this PR:

* use prepare for validation
* block changes after prepare has been run

jeenbroekstra commented on shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/ShaclSailConnection.java:
```diff
@@ -227,6 +233,28 @@ synchronized public void close() throws SailException {
+				valid = validate();
+
+				if (!valid) {
+					rollback();
```
I don't think you should call `rollback()` here. That's up to the user.

hmottestad replied:
What option does the user have? Does rollback do more than cancel the transaction and cleanup the store? If a user doesn’t rollback and just closes the store (and we don’t call rollback in close), is that bad in any way?

jeenbroekstra replied:
`close()` takes care of doing a `rollback()` if necessary, so that's taken care of. But the idea is that `prepare()` should not make things irreversible. For example, if `prepare()` fails not because of a SHACL validation issue but (for example) lack of disk space, a user should be able to do the following:

1. prepare -> exception (disk full) 
2. run a cleanup process (e.g. remove temp files)
3. prepare again -> success
4. commit

You can't do that if you choose to auto-invoke `rollback()`.

hmottestad replied:
That's a really good example @jeenbroekstra 

At the moment I ```rollback()``` inside ```commit()``` if the transaction fails. I guess that's wrong too, should probably defer that to ```close()``` as well.

jeenbroekstra commented on shacl/src/test/java/org/eclipse/rdf4j/sail/shacl/PrepareCommitTest.java:
```diff
@@ -0,0 +1,88 @@
+
+public class PrepareCommitTest {
+
+	@Test(expected = Throwable.class)
```
Would be good to be more specific about the exact expected exception.

jeenbroekstra replied:
And, for what it's worth: most base sail implementations allow further changes after a call to `prepare` - though I have yet to figure out if this is by design or simply a bug.

hmottestad replied:
I'm going to stick with throwing an exception if changes are made after a prepare(). It does say in the documentation that your only allowed to call prepare(), rollback(), commit() or potentially close() after prepare().

jeenbroekstra commented on shacl/src/test/java/org/eclipse/rdf4j/sail/shacl/PrepareCommitTest.java:
```diff
@@ -0,0 +1,88 @@
+package org.eclipse.rdf4j.sail.shacl;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2019 Eclipse RDF4J contributors.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Distribution License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/org/documents/edl-v10.php.
 *******************************************************************************/
 
package org.eclipse.rdf4j.sail.shacl;
```

hmottestad commented on compliance/store/src/test/java/org/eclipse/rdf4j/sail/memory/ShaclConcurrencyTest.java:
```diff
@@ -0,0 +1,35 @@
+	@Ignore
+	@Test
+	@Override
+	public void testConcurrentAddLargeTxnRollback()
```
It was actually ignored in the file I copied from :P

hmottestad commented:
@jeenbroekstra I think this is all done now and ready to merge.

I've removed automatic rollback from everything except close(). And added some tests to check that that works as intended.

I'm sticking with throwing an exception when you attempt to change the sail after calling prepare(), it makes sense according to the documentation and also makes validation a little bit simpler.

jeenbroekstra commented:
There appears to be a compilation failure in one of the shacl compliance tests on this PR, looks like it hasn't caught on that the shaclsail constructor changed.

hmottestad commented:
Looks like Jenkins might be stuck on Running org.eclipse.rdf4j.sail.shacl.ShaclIsolationLevelTest

That’s not very promising.

hmottestad commented:
Restarted Jenkins build. Same issue.
